components/script/dom/datatransferitem.rs

```diff
@@ -0,0 +1,153 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use dom_struct::dom_struct;
+
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::DataTransferItemBinding::{
+    DataTransferItemMethods, FunctionStringCallback,
+};
+use crate::dom::bindings::import::module::Rc;
+use crate::dom::bindings::reflector::{reflect_dom_object, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::bindings::weakref::MutableWeakRef;
+use crate::dom::datatransfer::DataTransfer;
+use crate::dom::file::File;
+use crate::dom::globalscope::GlobalScope;
+
+pub enum Kind<'a> {
+    Text(DOMString),
+    File(&'a File),
+}
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[crown::unrooted_must_root_lint::must_root]
+enum KindStorage {
+    Text(DOMString),
+    File(Dom<File>),
```

Comment:
As an example of how I think this has to be restructured:

- The file belongs on a `DragDataStore`, and does not have to be the dom object; rather, it could be a [`BlobImpl`](https://github.com/servo/servo/blob/672b37dd9c4b27fb9892ce7148d163444bf10de0/components/shared/script/serializable.rs#L65)
- As part of `as_file`, you would then create a new Dom `File`, using the data from the `BlobImpl`.